tests/test_process_action.cpp

```diff
@@ -0,0 +1,345 @@
+        observation.torque << 0.3, -0.25;
+
+        max_torque_Nm = 0.3;
+        safety_kd << 0.1, 0.1;
+
+        default_position_control_kp << 3, 4;
+        default_position_control_kd << .3, .4;
+    }
+};
+
+/**
+ * @brief Test if a valid torque command is returned unchanged (no safety).
+ */
+TEST_F(TestProcessDesiredAction, valid_torque_no_safety)
+{
+    // disable velocity damping for this test
+    safety_kd << 0., 0.;
+
+    Vector desired_torque;
+    desired_torque << 0.1, 0.2;
+    Types::Action action = Types::Action::Torque(desired_torque);
+
+    Types::Action resulting_action =
+        Functions::process_desired_action(action,
+                                          observation,
+                                          max_torque_Nm,
+                                          safety_kd,
+                                          default_position_control_kp,
+                                          default_position_control_kd);
+
+    ASSERT_EQ(0.1, resulting_action.torque[0]);
+    ASSERT_EQ(0.2, resulting_action.torque[1]);
+}
+
+/**
+ * @brief Test if clamping to max. torque is working.
+ */
+TEST_F(TestProcessDesiredAction, exceed_max_torque_no_safety)
+{
+    // disable velocity damping for this test
+    safety_kd << 0., 0.;
+
+    Vector desired_torque;
+    desired_torque << 0.5, -1.2;
+    Types::Action action = Types::Action::Torque(desired_torque);
+
+    Types::Action resulting_action =
+        Functions::process_desired_action(action,
+                                          observation,
+                                          max_torque_Nm,
+                                          safety_kd,
+                                          default_position_control_kp,
+                                          default_position_control_kd);
+
+    ASSERT_EQ(0.3, resulting_action.torque[0]);
+    ASSERT_EQ(-0.3, resulting_action.torque[1]);
+}
+
+/**
+ * @brief Test velocity damping with low velocity
+ */
+TEST_F(TestProcessDesiredAction, velocity_damping_low_velocity)
+{
+    Vector desired_torque;
+    desired_torque << 0.1, 0.2;
+    Types::Action action = Types::Action::Torque(desired_torque);
+
+    Types::Action resulting_action =
+        Functions::process_desired_action(action,
+                                          observation,
+                                          max_torque_Nm,
+                                          safety_kd,
+                                          default_position_control_kp,
+                                          default_position_control_kd);
+
+    // joint 0 has positive velocity so expecting a reduced torque.
+    // joint 1 has negative velocity so expecting an increased torque.
+    // both should remain within the max. torque range.
+    ASSERT_LT(resulting_action.torque[0], 0.1);
+    ASSERT_GE(resulting_action.torque[0], -0.3);
+    ASSERT_GT(resulting_action.torque[1], 0.2);
+    ASSERT_LE(resulting_action.torque[1], 0.3);
```

Comment:
also here it would be cleaner and more readable to use the variables. in fact i would probably put a for loop and check inside loop whether velocity for that joint is positive or negative and then do appropriate checks.

this way we are also robust to changes of values above. in fact it might make sense to run these tests which a bunch of different (random) values.

Reply:
I can see some value in randomized tests but on the other hand, making the tests work with random input would make them very complex and thus difficult to write/read.  It would also increase the risk of having bugs in the tests themselves.
Another thing is, that in my opinion unit tests should be deterministic, otherwise this can be quite annoying for CI.